### DISCUSSION TITLE
Return .single() for an inner join?

### DISCUSSION THREAD

**Discussion by jimmiejackson414** [supabase/supabase#7610] [Questions]
I have a query I'm running like this:

    const { data, error } = await supabase
      .from<User>('users')
      .select('*, customers!inner(*), subscriptions!inner(*)')
      .eq('id', userId)
      .single();
    console.log({ data });

This returns a User object, with a nested array of Customers and Subscriptions. Each User however should only have one Subscription and Customer object attached though. Is there a way to essentially call `.single()` on those inner joins?

EDIT: To clarify, I have a `customers` and `subscriptions` table, and each have a `user_id` column tied to `users`.

**Comment by steve-chavez** [comment]
Hm, so would you say that for your model, a user always has a single subscription and it's a single customer? `user` has a [one-to-one relationship](https://fmhelp.filemaker.com/help/18/fmp/en/index.html#page/FMP_Help/one-to-one-relationships.html) with `subscription` and `customer`?

**Reply by jimmiejackson414** [reply]
Correct, in my app a subscription and customer are both Stripe-related tables, and in my scenario a user should only ever have one Customer record and one Subscription record tied to that User.

**Reply by steve-chavez** [reply]
I see. Right now you'll have to work around this on the client side but this will be detected automatically once  https://github.com/supabase/postgrest-js/issues/223 is finished(soonish).

**Reply by jimmiejackson414** [reply]
Ah very cool, Ill be sure to follow along with its progress. Thanks!

**Reply by Kamahl19** [reply]
@steve-chavez should this work automatically now? I have this query
```
await supabase
        .from('books')
        .select(
          `
          *,
          reading_list(*)
        `,
        )
        .eq('id', bookId)
        .maybeSingle();
```
but it always returns `reading_list` as array `reading_list[0]`.

My schema is
```
create table "public"."books" (
    "id" uuid not null default gen_random_uuid(),
    "title" text not null,
    "author" text not null,
    "description" text not null,
    "createdAt" timestamp with time zone not null default now()
);

create unique index "books_pkey" on "public"."books" using btree ("id");

alter table "public"."books" add constraint "books_pkey" primary key using index "books_pkey";

create table "public"."reading_list" (
    "userId" uuid not null,
    "bookId" uuid not null,
    "finished" boolean not null default false,
    "rating" smallint not null default '0',
    "note" text not null default '',
    "createdAt" timestamp with time zone not null default now()
);

create unique index "reading_list_pkey" on "public"."reading_list" using btree ("userId", "bookId");

alter table "public"."reading_list" add constraint "reading_list_pkey" primary key using index "reading_list_pkey";
alter table "public"."reading_list" add constraint "reading_list_bookId_fkey" foreign key ("bookId") references books("id") on delete cascade not valid;
alter table "public"."reading_list" add constraint "reading_list_userId_fkey" foreign key ("userId") references auth.users("id") on delete cascade not valid;

alter table "public"."reading_list" validate constraint "reading_list_bookId_fkey";
alter table "public"."reading_list" validate constraint "reading_list_userId_fkey";
```

**Reply by steve-chavez** [reply]
@Kamahl19 Should work when the relationship is one-to-one, see https://postgrest.org/en/stable/references/api/resource_embedding.html#one-to-one-relationships

**Reply by Kamahl19** [reply]
"reading_list" has 2 FKs, one pointing to the "books" table and another to "users" table. They together create a unique index. There is a 1:1 relationship. Is my usecase not supported?

**Reply by steve-chavez** [reply]
@Kamahl19 Can you format your snippet so it's all declarative (no ALTER TABLE, instead declare the FKs with REFERENCES and inline PK), otherwise it's hard to visualize.

**Reply by steve-chavez** [reply]
Also note that you can define one-to-one relationships manually, see https://postgrest.org/en/stable/references/api/resource_embedding.html#computed-relationships

**Reply by Kamahl19** [reply]
```sql
CREATE TABLE
  "books" (
    "id" UUID NOT NULL DEFAULT gen_random_uuid(),
    "title" TEXT NOT NULL,
    "author" TEXT NOT NULL,
    "description" TEXT NOT NULL,
    "createdAt" TIMESTAMP WITH TIME ZONE NOT NULL DEFAULT now(),
    CONSTRAINT "books_pkey" PRIMARY KEY ("id")
  );

CREATE TABLE
  "reading_list" (
    "userId" UUID NOT NULL,
    "bookId" UUID NOT NULL,
    "finished" BOOLEAN NOT NULL DEFAULT false,
    "rating" SMALLINT NOT NULL DEFAULT '0'::smallint,
    "note" TEXT NOT NULL DEFAULT ''::text,
    "createdAt" TIMESTAMP WITH TIME ZONE NOT NULL DEFAULT now(),
    CONSTRAINT "reading_list_pkey" PRIMARY KEY ("userId", "bookId"),
    CONSTRAINT "reading_list_userId_fkey" FOREIGN KEY ("userId") REFERENCES "auth"."users" ("id") ON DELETE CASCADE,
    CONSTRAINT "reading_list_bookId_fkey" FOREIGN KEY ("bookId") REFERENCES "books" ("id") ON DELETE CASCADE
  );
```

**Reply by steve-chavez** [reply]
Hm, I don't see the one-to-one according to your table structure. If anything it looks like you have a many-to-many on `books` and `users` using the `reading_list` as join table. Check https://postgrest.org/en/stable/references/api/resource_embedding.html#many-to-many-relationships

**Reply by steve-chavez** [reply]
`books` and `reading_list` should form a many-to-one though. Maybe try doing `NOTIFY pgrst, 'reload schema'`, in case the schema cache didn't get updated somehow.

**Comment by RyannGalea** [comment]
Any progress on this? I'm in the same situation where the inner join for a parent should always be singular.

**Reply by madeleineostoja** [reply]
This has been patched in PostgREST, supabase just needs a version bump. It’s a breaking change so I assume it’ll be rolled into the v2 RC

**Reply by activenode** [reply]
Is there a commit/patch link?

**Comment by chirastefan** [comment]
is there a plan to implement this in supabase_flutter also?

**Reply by steve-chavez** [reply]
Should work transparently for flutter, new projects already come with this change. See https://github.com/supabase/postgrest-js/issues/223.

**Comment by gabrielsestrem** [comment]
@steve-chavez have you guys published something related to this recently? After adding a Small add-on to my plan, my inner joins are now returning {} instead of an array [{}].

That change broke my entire application.

**Reply by gabrielsestrem** [reply]
nevermind... I already changed all the code to the new format

**Reply by steve-chavez** [reply]
@gabrielsestrem Hm, sounds this is related to this newish feature https://github.com/supabase/postgrest-js/issues/223

**Comment by nlarusstone** [comment]
Has anyone gotten the correct typescript types to be generated here? I'm on the new version of Postgrest (10.1.2) but typescript still thinks this could be either a list or a single item (even though it always returns just a single object)

**Reply by Rykuno** [reply]
> Has anyone gotten the correct typescript types to be generated here? I'm on the new version of Postgrest (10.1.2) but typescript still thinks this could be either a list or a single item (even though it always returns just a single object)

Same. This is a very annoying issue :/

**Reply by RyannGalea** [reply]
In most cases I'll just create my own types to inject the data how it should be seen in the app, in this example I Omit the 'last_message' field as it thinks it's a string as it's a JSONB field.

I then type it along with some other joined data.

```
type BaseConversation    = Omit<Database['public']['Tables'][Tables.conversations]['Row'], 'last_message'>
export type Conversation = BaseConversation & {
  contact?: Partial<Contact>,
  user?: Partial<User>,
  last_message?: { body?: string, dateCreated?: string | Date }
}
```

**Comment by hipdev** [comment]
What is working in my case is: 
if is a one to one relation I give the unique option in the dashboard, then after I just generate types again. The isOneToOne option comes as true. The type is a mess but is working if I use  the foreignKeyName value in the supabase generated types in the join query.


![Screenshot 2023-11-08 at 9 48 29 PM](https://github.com/supabase/supabase/assets/48492084/e4ae4a2b-261a-421a-beac-ca4a14563d22)

**Reply by albertonii** [reply]
@hipdev this is true, in my case I can avoid the Linter error if I do this in my schema types for my app in NextJS created by Supabase with the script:     "npx supabase gen types typescript --db-url [DB_URL] > ./src/lib/schema.ts ".

But this is extremely hard to maintain, everytime types are generated I need to come to find the FK and insert that isOneToOne, also... I have multiple FK to update.

Anyways thanks for the temporary solution.

**Reply by logemann** [reply]
Weird. The manual isOneToOne  workararound doesnt even work for me. Still the same linter errors.

**Reply by hipdev** [reply]
@logemann try it using this approach:
Create the relation with the foreign key name
<img width="601" alt="image" src="https://github.com/supabase/supabase/assets/48492084/2b4aef32-b2a7-4539-a947-c71b92d45319">

Seems to resolve the issue in the meantime

**Reply by logemann** [reply]
I fixed it by upgrading to latest supabase-js npm package ;-)

**Reply by hipdev** [reply]
So, the Supa squad fixed it 😄

**Reply by whollacsek** [reply]
I confirm switching to `foreignKeyName` works but we should be able to use column name as well, still not an ideal DX

**Reply by louisprp** [reply]
For me it works as it's supposed to work, picking up all relations as they are intended, while also dynamically providing the correct type from the complex query to reuse.

Having a function defined like this

```
export const getOrders = async (
  client: SupabaseClient<Database>
) => {
  const query = client
    .from("orders")
    .select(
      "id, customer_id:users!orders_customer_id_fkey(*), supplier_id:users!orders_supplier_id_fkey(*), status, created_at, orders_products (*)"
    )
  return query
}
```

I can now get the type of a single order like this:

`export type Order = NonNullable<AsyncReturnType<typeof getOrders>["data"]>[number]`

Where `AsyncReturnType` looks like this:

```
// Types for getting the return type of an asynchronous function
type PromiseType<T> = T extends Promise<infer U> ? U : never

export type ErrorType = { error: string }

/**
 * Represents the return type of an asynchronous function.
 * It extracts the resolved value from a Promise and excludes any potential error type.
 * Used primarily to get supabase action return types.
 */
export type AsyncReturnType<T extends (...args: any) => Promise<any>> = Exclude<
  PromiseType<ReturnType<T>>,
  ErrorType
>
```

**Comment by neoromantic** [comment]
I still face same problem and couldn't find solution online.

Looks like this

```js
  const query = supabase
    .from("transactions")
    .select(`
                amount,
                datetime,
                type,
                counterparty:counterparties!counterparty_id (
                    id,
                    name
                )
            `)
```

<img width="959" height="171" alt="image" src="https://github.com/user-attachments/assets/499aa89c-0eaa-4b79-942f-12923a09e2e1" />

So, the types generated expect the result to have an array of counterparties, but this is a foreign key to singular item. 

Any hope to fix it short of casting to my own type?
